### PR TITLE
Update c5 ingester with new scheduler jobs

### PR DESCRIPTION
**Pull request by AdamDM**

### CODE REVIEW

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java [hunk 1]
```diff
@@ -40,24 +40,41 @@
     
     @PostConstruct
     public void startBackgroundTasks() {
-        scheduler.schedule(fiveUpdater().withName("Five Updater"), DAILY);
+        scheduler.schedule(
+                fiveUpdater(FiveUpdater.TimeFrame.ALL)
+                        .withName("Five Updater")
+                , DAILY
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: Let's change this one to 1/week on Friday night

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java [hunk 2]
```diff
@@ -40,24 +40,41 @@
...
+        scheduler.schedule(
+                fiveUpdater(FiveUpdater.TimeFrame.TODAY)
+                        .withName("Five Today Updater")
+                , RepetitionRules.NEVER
+        );
+        log.info("Installed FiveToday updater");
+        scheduler.schedule(
+                fiveUpdater(FiveUpdater.TimeFrame.PLUS_MINUS_7_DAYS)
+                        .withName("Five +-7 Days Updater")
+                , RepetitionRules.NEVER
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: And this one to 1/day

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java [hunk 3]
```diff
@@ -40,24 +40,41 @@
...
-                fiveLocationPolicyIds(),
-                Configurer.get("five.timeout.socket", "180").toInt()
-        );
+    public FiveUpdater fiveUpdater(FiveUpdater.TimeFrame timeFrame) {
+        return FiveUpdater.builder()
+                .withContentWriter(contentWriter)
+                .withChannelResolver(channelResolver)
+                .withContentResolver(contentResolver)
+                .withLocationPolicyIds(fiveLocationPolicyIds())
+                .withSocketTimeout(Configurer.get("five.timeout.socket", "180").toInt())
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: Let's move this parameter to the top of the class as a field to be consistent with how configuration works elsewhere

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java (hunk 2 above)
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: put these commas on the line above

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java (hunk 3 above)
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: Extract the Configurer.get to the top of the class with checkNotNull

File: src/test/java/org/atlasapi/remotesite/five/FiveUpdaterTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        assertEquals(expectedCall, foundCall);
+    }
+
+    private void getAndTestApiCall(FiveUpdater.TimeFrame timeFrame, String startDateTime, String endDateTime) {
+        FiveUpdater fiveUpdater = fiveUpdaterBuilder
+                .withTimeFrame(timeFrame)
+                .build();
+
+        DateTimeFormatter formatter = DateTimeFormat.forPattern("y-M-d'T'hh:mm:ss'Z'");
```
Comment (Dalamar42) [inline src/test/java/org/atlasapi/remotesite/five/FiveUpdaterTest.java]: This is ISO 8601. There is already a predefined formatter for this

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 5]
```diff
@@ -153,11 +184,32 @@ private SimpleHttpClient buildFetcher() {
                 .build();
     }
 
-    private String getApiCall(int startingPoint) {
-        return String.format("%s/shows?offset=%d&limit=%d", BASE_API_URL, startingPoint, LIMIT);
+    @VisibleForTesting
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: This is part of the class' internal logic. We should not be exposing it for testing. If the class does too many things to test we should break up the class

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 6]
```diff
@@ -153,11 +184,32 @@ private SimpleHttpClient buildFetcher() {
...
+    public String getApiCall(int startingPoint, String channelId, DateTime dateTime) {
+        DateTimeFormatter formatter = DateTimeFormat.forPattern("y-M-d'T'hh:mm:ss'Z'");
+        String baseUri = "https://pdb.five.tv/internal/channels/%s/transmissions"
+                + "?expand=watchable&from_time=%s&until_time=%s";
+        if (timeFrame == TimeFrame.TODAY) {
+            DateTime todayStartDateTime = dateTime.withTimeAtStartOfDay();
+            DateTime todayEndDateTime = todayStartDateTime
+                    .plusDays(1)
+                    .withTimeAtStartOfDay()
+                    .minusMinutes(1);
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: Why `.minusMinutes(1)` and not `.minusMillis(1)`?

Also if their API considers to the `to` date to be exclusive (needs checking) then you don't need to subtract anything. You can just set the `to` as the next midnight

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 7]
```diff
@@ -132,13 +114,62 @@ public void runTask() {
         }
     }
 
+    private ArrayList<Element> combineElements(int startingPoint) throws Exception {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: Return immutable collections and if you absolutely need a mutable one which you don't need here then define them using the interface `List<>` not the concrete type `ArrayList<>`

Comment (Dalamar42) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: This method is hard to follow. Let's speak f2f.

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 8]
```diff
@@ -132,13 +114,62 @@ public void runTask() {
...
+            }
+
+            return elementArrayList;
+        }
+
+        // get list of channel ids
+        Document channelIdsDocument = streamHttpClient.get(new SimpleHttpRequest<>(
+                "https://pdb.five.tv/internal/channels",
+                (responsePrologue, inputStream) -> parseResponse(inputStream)
+        ));
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: a "combineElements" method should probably just combine elements, not make a http call

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 9]
```diff
@@ -132,13 +114,62 @@ public void runTask() {
...
+        Elements channelIds = channelIdsDocument.getRootElement().getChildElements("channels");
+
+        for (int i = 0; i < channelIds.size(); i++) {
+            Element id = channelIds.get(i).getFirstChildElement("id");
+            Elements showElements = getElements(startingPoint, id.getValue());
+            for (int j = 0; j < showElements.size(); j++) {
+                Element element = showElements.get(i);
+                elementArrayList.add(element);
+            }
+        }
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: Pretty sure this can be done cleaner using a lambda

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java [hunk 10]
```diff
@@ -153,11 +184,32 @@ private SimpleHttpClient buildFetcher() {
...
+            DateTime todayStartDateTime = dateTime.withTimeAtStartOfDay();
+            DateTime todayEndDateTime = todayStartDateTime
+                    .plusDays(1)
+                    .withTimeAtStartOfDay()
+                    .minusMinutes(1);
+            String todayStart = todayStartDateTime.toString(formatter);
+            String todayFinish = todayEndDateTime.toString(formatter);
+            return String.format(baseUri, channelId, todayStart, todayFinish);
+        } else if (timeFrame == TimeFrame.PLUS_MINUS_7_DAYS) {
+            DateTime weekAgo = dateTime.minusDays(7);
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: I'd get rid of the assignment to a DateTime and just add .toString(formatter) on the end of this to just set a String seeing as you don't use the DateTime anywhere else

File: src/main/java/org/atlasapi/remotesite/five/FiveModule.java [hunk 11]
```diff
@@ -37,27 +39,45 @@
...
-    
+    private final int socketTimeout = checkNotNull(Configurer.get("five.timeout.socket", "180").toInt());
+
     @PostConstruct
     public void startBackgroundTasks() {
-        scheduler.schedule(fiveUpdater().withName("Five Updater"), DAILY);
+        scheduler.schedule(
+                fiveUpdater(FiveUpdater.TimeFrame.ALL)
+                        .withName("Five Updater"),
+                DAILY
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/remotesite/five/FiveModule.java]: Inconsistent with the other RepetitionRules. Having it as RepetitionRule.X makes it clearer what this is

File: src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java (hunk 6 above)
Comment (AdamDM) [inline src/main/java/org/atlasapi/remotesite/five/FiveUpdater.java]: it was specified by Romain this way

File: src/test/java/org/atlasapi/remotesite/five/FiveUpdaterTest.java (hunk 4 above)
Comment (AdamDM) [inline src/test/java/org/atlasapi/remotesite/five/FiveUpdaterTest.java]: The predefined formatter seems to give a +1.00 for the time zone

File: Dockerfile [hunk 12]
```diff
@@ -10,6 +10,7 @@ ENV SERVER_PORT="8080" \
     EQUIV_STREAM_UPDATER_ENABLED="false" \
     EQUIV_UPDATER_ENABLED="false" \
     IDS_GENERATE="false" \
+    JSSE_ENABLESNIEXTENSION="false" \
```
Comment (jamiesprax) [inline Dockerfile]: Keep the naming the same. "JSSE_SNI_EXTENSION_ENABLED" or something, I'm not quite sure where the different parts end
